### PR TITLE
CTPPS pixel clusterizer: call calibration by reference and use of map instead of set

### PR DESCRIPTION
**Pull request by fabferro**
Removing a copy in the call to gain calibration and use of map instead of set to speed up execution

### CODE REVIEW

**Comment by slava77** [comment]
since we started touching this,
there is another place where CPU use could be reduced:
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_9_4_X_2017-09-24-0000-sign962.294736.reCTPPS.pp/162
points to very high cost of iterating over calib_rpix_digi_set_.
If the set can be changed to a map with a key column*maxRow+row, lookup should be much easier than a linear scan over the full set every time a pixel is needed.
another alternative is to fill a vector size maxRow*maxCol, but this is perhaps too costly on low occupancy.

**Comment by fabferro** [comment]
@slava77 I've implemented the use of map instead of set. Please check at your convinience that this speeds up execution.

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
@smuzaffar 
Something is odd in this PR history:

The last test by jenkins is "Tested at: 284575f", but it does not appear in the normal way on this PR page (the last visible commit is 2475d19 ).
The head of the topic branch is indeed this commit
https://github.com/CTPPS/cmssw/tree/ctpps_pixel_cluster_patch2

Is it something on the github end that has hidden the last commit or did the bot edit something to make it disappear?

**Comment by slava77** [comment]
@cmsbuild please test

reco comparisons should now include CTPPS pixel data

**Comment by slava77** [comment]
@fabferro 
I observe some minor differences from this PR (the plots are from loca reco on a skim from ZeroBias2 run 294736 with events with at least 2K pixel digis)

number of clusters per det
all_sign970vsorig_ctppsfromaod294736c_ctppspixelclusteredmdetsetvector_ctppspixelclusters__rectpps_obj__sets_dataat_size

cluster size
all_sign970vsorig_ctppsfromaod294736c_ctppspixelclusteredmdetsetvector_ctppspixelclusters__rectpps_obj__sets_data_size
all_sign970vsorig_ctppsfromaod294736c_ctppspixelclusteredmdetsetvector_ctppspixelclusters__rectpps_obj__sets_data_sizecol

It seems like something is happening on the edge in x:
all_sign970vsorig_ctppsfromaod294736c_ctppspixelrechitedmdetsetvector_ctppspixelrechits__rectpps_obj__sets_data_getpoint_x
all_sign970vsorig_ctppsfromaod294736c_ctppspixelrechitedmdetsetvector_ctppspixelrechits__rectpps_obj__sets_data_minpixelrow

while y seems OK (no localization)
all_sign970vsorig_ctppsfromaod294736c_ctppspixelrechitedmdetsetvector_ctppspixelrechits__rectpps_obj__sets_data_getpoint_y


There are no changes in the tracks.

Perhaps there is something off in the indexing (either on the edge cases or because there are multiple digis at the same row,col which would be lost with std::map).
Please check.

**Comment by fabferro** [comment]
@slava77 
I think that the diffrence you observe are due to a buggy management of the pointers during the clusterization done with the set, probably here https://github.com/cms-sw/cmssw/pull/20642/files#diff-c0803683db8122e1cb16f90ed651a837L110
In fact, I observe that using the map I recover pixels that were "lost" with the sets and the cluster size increases (see the picture: magenta=new). This is propagated to rechits and tracks but it has very poor effect on them. 
So, in summary, I think we have fixed a bug we were not aware of.
canvas_1

**Comment by slava77** [comment]
On 10/3/17 2:13 PM, Fabrizio Ferro wrote:
> So, in summary, I think we have fixed a bug we were not aware of.

nice

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20642	284575f
- initially, to speed up clustering; in the end it appears that also some subtle bug was fixed as well
- jenkins tests pass and comparisons show no differences
- local tests on a set of events with high occupancy shows that the timing of ctppsPixelClusters goes down by a factor of about 20 from 0.11 s/event to 3.6 ms/event; comparisons here show some differences as posted in https://github.com/cms-sw/cmssw/pull/20642#issuecomment-333933646 these appear to be an improvement (bugfix) in the code.

File: RecoCTPPS/PixelLocal/src/RPixDetClusterizer.cc [hunk 1]
```diff
@@ -65,29 +65,31 @@ void RPixDetClusterizer::buildClusters(unsigned int detId, const std::vector<CTP
...
   for(auto SeedIt : SeedVector_){
     make_cluster( SeedIt, clusters);
   }
 }
 
 void RPixDetClusterizer::make_cluster(RPixCalibDigi const &aSeed,  std::vector<CTPPSPixelCluster> &clusters ){
 /// check if seed already used
-  if(calib_rpix_digi_set_.size()==0 || calib_rpix_digi_set_.find(aSeed)==calib_rpix_digi_set_.end() ){
+  unsigned int seedIndex = aSeed.column()*maxRow + aSeed.row();
+  if(calib_rpix_digi_map_.empty() || calib_rpix_digi_map_.find(seedIndex)==calib_rpix_digi_map_.end() ){
```
Comment (davidlange6) [inline RecoCTPPS/PixelLocal/src/RPixDetClusterizer.cc]: hi @fabferro - the check on .empty() is no longer needed here. (maybe it never was)

Comment (fabferro) [reply RecoCTPPS/PixelLocal/src/RPixDetClusterizer.cc]: Yes, "find" should do the job already

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20642	b7fe46b
- technical update since the last signoff by reco https://github.com/cms-sw/cmssw/pull/20642/files/284575feb45b3521371d64ff9597d3c487655333..b7fe46b741a7f645c65325ca7db3007ba704bd31
- jenkins tests pass

**Comment by davidlange6** [comment]
+1

**Comment by slava77** [comment]
@fabferro 
please update the PR title and description to reflect what is done in the PR
